Gist shortcode: bring in sync with WordPress.com and prepare for Fusion

jeherve opened the pull request:
#### Changes proposed in this Pull Request:

The Gist shortcode that ships with WordPress.com is quite different from the one in Jetpack. The 2 files were never kept in sync, and have diverged a lot over the years, to the point where they don't support the same range of formats.

To corrct this, I had to make a few changes to the shortcode:

- Extract all id sanitization in separate function, `jetpack_gist_get_shortcode_id`
- Strip username from gist ids. They are not necessary, and break the AMP implementation.

In addition to this, I fixed all phpcs warnings.

On  WordPress.com: D27151-code

#### Testing instructions:

* Add the following to a classic block in the editor:

```
<p>Supported formats:</p>
<p>Full URL</p>
<p>https://gist.github.com/57cc50246aab776e110060926a2face2</p>
<p>Full URL with username:</p>
<p>https://gist.github.com/jeherve/57cc50246aab776e110060926a2face2</p>
<p>Full URL linking to specific file:</p>
<p>https://gist.github.com/jeherve/57cc50246aab776e110060926a2face2#file-wp-config-php</p>
<p>Full URL, no username, linking to specific file:</p>
<p>https://gist.github.com/57cc50246aab776e110060926a2face2#file-wp-config-php</p>
<p>Gist ID:</p>
<p>[gist]57cc50246aab776e110060926a2face2[/gist]</p>
<p>Gist ID within tag:</p>
<p>[gist 57cc50246aab776e110060926a2face2]</p>
<p>Gist ID with username:</p>
<p>[gist jeherve/57cc50246aab776e110060926a2face2]</p>
<p>Gist private ID with username:</p>
<p>[gist xknown/fc5891af153e2cf365c9]</p>
<p>Invalid format:</p>
<p>[gist xknown/fc5891af153e2cf365c9/raw?]</p>
```

* All shortcodes should work, except for the last one.

#### Proposed changelog entry for your changes:

* Shortcodes: support a wider range of Gist embeds.

jeherve committed:
Shortcodes: fix all phpcs warnings in preparation for Fusion

jeherve committed:
Shortcodes: improve validation of Gist embeds.

Merges r95254-wpcom

- Avoid trying to embed gists when the URL is not from gist.github.com.
- Avoid trying to embed gists with invalid characters in IDs.

jeherve committed:
Shortcodes: expand unit test coverage for Gist IDs.

See 8fc3996c2fc2ff57df3751f3a3e320f6d5feed5d

jeherve committed:
Extract all id sanitization in separate function

- This makes for easier maintenance.
- It allows us to do most sanitization only once, instead of having to do it inside the AMP check as well.
- I opted to strip username from gist ids. They are not necessary, and break the AMP implementation.

jeherve committed:
Gist shortcode: new and updated tests.

Existing tests are not updated to take into account that we strip GitHub usernames.
New tests cover new use cases, so all potential gist shortcode formats are covered.

jeherve committed:
Add supported formats at the top of the file

simison commented on modules/shortcodes/gist.php:
```diff
@@ -128,14 +201,13 @@ function github_gist_shortcode( $atts, $content = '' ) {
 function github_gist_simple_embed( $id ) {
 	$id = str_replace( 'json', 'js', $id );
-	return '<script type="text/javascript" src="https://gist.github.com/' . $id . '"></script>';
+	return '<script type="text/javascript" src="https://gist.github.com/' . $id . '"></script>'; // phpcs:ignore WordPress.WP.EnqueuedResources.NonEnqueuedScript
```
`type` in `<script>` tags is HTML4. 🤔 Maybe drop it while at it?

jeherve replied:
👍 Done in d855731b31dc20cb844b8d55d443814f21485eee

jeherve committed:
Remove type from script tag.

See https://github.com/Automattic/jetpack/pull/12096#discussion_r276922219

jeherve committed:
Extend tests to match all possible formats.

mdawaffe commented on modules/shortcodes/gist.php:
```diff
@@ -22,11 +34,73 @@
+	}
+
+	// Not a URL nor an ID? Only keep one of "username/id", "/username/id", "id".
+	if ( preg_match( '#^/?(([a-z0-9-_]+/)?([a-z0-9]+))$#i', $gist, $matches ) ) {
```
It doesn't matter in this regex, but it's usually best to put any `-` that is not part of a range at the end of the character class so that it cannot be interpreted as a range character.

```suggestion
	if ( preg_match( '#^/?(([a-z0-9_-]+/)?([a-z0-9]+))$#i', $gist, $matches ) ) {
```

mdawaffe commented on modules/shortcodes/gist.php:
```diff
@@ -22,11 +34,73 @@
+
+	// Not a URL nor an ID? Only keep one of "username/id", "/username/id", "id".
+	if ( preg_match( '#^/?(([a-z0-9-_]+/)?([a-z0-9]+))$#i', $gist, $matches ) ) {
+		$gist_info['id'] = $matches[1];
```
Instead of doing a second `preg_replace()` a few lines down ("If there is one, strip…"), can we pull what we need from this first `preg_replace()`?

```suggestion
		$gist_info['id'] = $matches[3];
```

jeherve replied:
Good call. Done in 7bf2569306628400c82f895d6de99ec8cf3d7def

mdawaffe commented on modules/shortcodes/gist.php:
```diff
@@ -45,16 +119,34 @@ function github_gist_shortcode( $atts, $content = '' ) {
+	// Parse a URL to get an ID we can use.
+	$gist_info = jetpack_gist_get_shortcode_id( $id );
+	if ( empty( $gist_info['id'] ) ) {
+		return '<!-- Invalid Gist ID -->';
```
Do we ever show errors to users who can edit the post for shortcodes? I know we do for some widgets.

jeherve replied:
That's a good point. Done in 54f5d763e11ce80f6576aa26991554cf1dfa73fa

mdawaffe commented on modules/shortcodes/gist.php:
```diff
@@ -45,16 +119,34 @@ function github_gist_shortcode( $atts, $content = '' ) {
+
+	// Replace - by . to get a real file name from slug.
+	if ( ! empty( $file ) ) {
+		$file = preg_replace( '/\-(?!.*\-)/', '.', $file );
```
In theory, `$file` could be very long (if it comes from `$atts['file']`. `strrpos()`/`substr_replace()` might be more performant on very long strings with lots of `-`s.

jeherve replied:
👍 Done in ff8874147e370f5060d59758cb176c53cf9b2c98

mdawaffe commented on modules/shortcodes/gist.php:
```diff
@@ -22,11 +34,73 @@
+
+		// Keep the unique identifier without any leading or trailing slashes.
+		if ( ! empty( $parsed_url['path'] ) ) {
+			$gist_info['id'] = preg_replace( '/(?:\/)([^\.]+)\./', '$1', $parsed_url['path'] );
```
Does `$parsed_url['path']` always start with a `/`? If so, we should anchor the regex with `^`.

We don't need to group the `/`.
```suggestion
			$gist_info['id'] = preg_replace( '/^\/([^\.]+)\./', '$1', $parsed_url['path'] );
```

mdawaffe commented on modules/shortcodes/gist.php:
```diff
@@ -22,11 +34,73 @@
+
+		// Keep the file name if there was one.
+		if ( ! empty( $parsed_url['fragment'] ) ) {
+			$gist_info['file'] = preg_replace( '/(?:file-)(.+)/', '$1', $parsed_url['fragment'] );
```
Do we need to check that the #fragment has this form? Or can we trust the embed JS to handle any malformed data?

jeherve replied:
I mostly used this to remove the `file-` prefix from the fragment. Do you think I should use `substr` instead?

mdawaffe replied:
I think it's a fine way to get rid of `file-`. Looks like the embed js 404s if the file is invalid (currently, at least), so it's probably fine.

mdawaffe commented on modules/shortcodes/gist.php:
```diff
@@ -128,14 +201,13 @@ function github_gist_shortcode( $atts, $content = '' ) {
 function github_gist_simple_embed( $id ) {
 	$id = str_replace( 'json', 'js', $id );
-	return '<script type="text/javascript" src="https://gist.github.com/' . $id . '"></script>';
+	return '<script src="https://gist.github.com/' . $id . '"></script>'; // phpcs:ignore WordPress.WP.EnqueuedResources.NonEnqueuedScript
```
Let's escape this so that no future changes on lines above can cause problems.
```suggestion
	return '<script src="' . esc_url( "https://gist.github.com/$id" ) . '"></script>'; // phpcs:ignore WordPress.WP.EnqueuedResources.NonEnqueuedScript
```

mdawaffe reviewed (commented):
I left some nits and suggestions inline. I didn't test any of the suggestions :)

I'm surprised we need to support all these formats. Nice work writing tests for them all!

mdawaffe committed:
Escape gist URL outputted in editor

Co-Authored-By: jeherve <jeremy@tagada.hu>

mdawaffe committed:
Simplify regex

it's usually best to put any - that is not part of a range at the end of the character class so that it cannot be interpreted as a range character.

Co-Authored-By: jeherve <jeremy@tagada.hu>

jeherve committed:
Simplify how we only keep the ID from a gist URL

See https://github.com/Automattic/jetpack/pull/12096#discussion_r277038467

jeherve committed:
Use strrpos and substr_replace instead of a regex

See https://github.com/Automattic/jetpack/pull/12096/files#r277039431

jeherve committed:
Display message to post authors when a gist can't be displayed

See https://github.com/Automattic/jetpack/pull/12096#discussion_r277038762

mdawaffe committed:
Regex improvements

See https://github.com/Automattic/jetpack/pull/12096#discussion_r277040103

Co-Authored-By: jeherve <jeremy@tagada.hu>

kraftbj reviewed (approved):
Nice. Confirmed each expected Gist format renders.
